backend/assignments/2/src/main/java/com/example/assignment2/service/GeoCodingImp.java

```diff
@@ -0,0 +1,124 @@
+            return cacheResponse;
+        }
+        try{
+           URI uri = uriMaker.requestURLBuilder(address);
+            log.info(String.valueOf(uri));
+            ResponseEntity<String> aPIresponse = restTemplate
+                    .exchange(
+                            uri,
+                            HttpMethod.GET,new HttpEntity<>(null),String.class);
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode root = mapper.readTree(aPIresponse.getBody());
+            if (root.has("error")) {
+                throw new JsonResponseNodeException("error in Json Node");
+            }
+            JsonNode firstResult = root.path("data").path(0);
+            if (firstResult.isMissingNode()) {
+                throw new JsonResponseNodeException("missing node in Json");
+            }
+            Double latitude = firstResult.path("latitude").asDouble();
+            Double longitude = firstResult.path("longitude").asDouble();
+
+            GeoCodingResponseDTO responseDTO = GeoCodingResponseDTO.builder()
+                    .latitude(latitude)
+                    .longitude(longitude)
+                    .build();
+            if(!address.contains("goa")) {
+                Objects.requireNonNull(cacheManager.getCache("geocoding")).put(address, responseDTO);
+            }
+            return responseDTO;
+
+        } catch (Exception e){
+            log.error("Something went wrong while getting value from positionstack api"+ e.getMessage());
+            throw new InvalidInputException(e.getMessage());
+        }
+    }
+
+    public ReverseGeoCodingResponseDTO getReverseGeoCoding(ReverseGeoCodingRequestDTO requestDTO){
+        try{
+        double latitude = requestDTO.getLatitude();
+        double longitude = requestDTO.getLongitude();
+        String cacheKey = latitude+ ","+longitude;
+        log.info(cacheKey);
+        if (latitude < -90 || latitude > 90 || longitude < -180 || longitude > 180) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,"invalid latitude or longitude");
+        }
+
+        ReverseGeoCodingResponseDTO cacheResponse = Objects.requireNonNull(cacheManager.getCache("reverse-geocoding")).get(cacheKey,ReverseGeoCodingResponseDTO.class);
+        log.info("Cache search response:"+ cacheResponse);
+        if(cacheResponse != null){
+            return cacheResponse;
+        }
+
+            URI uri = uriMaker.requestURLBuilder(latitude,longitude);
+            log.debug(String.valueOf(uri));
+            ResponseEntity<String> response = restTemplate
+                    .exchange(
+                            uri,
+                            HttpMethod.GET,new HttpEntity<>(null),String.class);
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode root = mapper.readTree(response.getBody());
+            if (root.has("error")) {
+                throw new JsonResponseNodeException("error in Json Node");
```

Comment:
error messages / exceptions must be stored in constants, also the messages must be intuitive